lib/livebook/zta/livebook_teams.ex

```diff
@@ -0,0 +1,147 @@
+defmodule Livebook.ZTA.LivebookTeams do
+  use GenServer
+  use LivebookWeb, :verified_routes
+
+  require Logger
+
+  alias Livebook.Teams
+
+  import Plug.Conn
+  import Phoenix.Controller
+
+  @behaviour Livebook.ZTA
+
+  @impl true
+  def child_spec(opts) do
+    %{id: __MODULE__, start: {__MODULE__, :start_link, [opts]}}
+  end
+
+  def start_link(opts) do
+    name = Keyword.fetch!(opts, :name)
+
+    if id = Livebook.Config.teams_auth_hub_id() do
+      team = Livebook.Hubs.fetch_hub!(id)
+      Livebook.ZTA.put(name, team)
+
+      :ignore
+    end
+  end
+
+  @impl true
+  def init(arg) do
+    {:ok, arg}
+  end
+
+  @impl true
+  def authenticate(name, conn, _opts) do
+    team = Livebook.ZTA.get(name)
+    auth = Livebook.Config.teams_auth()
+
+    if Livebook.Hubs.TeamClient.livebook_teams_zta?(team.id) do
+      handle_request(conn, team, auth, conn.params)
+    else
+      {conn, nil}
+    end
+  end
+
+  defp handle_request(conn, team, :online, %{"teams_identity" => _, "code" => code}) do
+    with {:ok, access_token} <- retrieve_access_token(team, code),
+         {:ok, metadata} <- get_user_info(team, access_token) do
+      {conn
+       |> put_session(:identity_data, metadata)
+       |> redirect(to: conn.request_path)
+       |> halt(), metadata}
+    else
+      _ ->
+        {conn
+         |> redirect(to: conn.request_path)
+         |> put_session(:teams_error, true)
+         |> halt(), nil}
+    end
+  end
+
+  defp handle_request(conn, _team, :online, %{"teams_identity" => _, "failed_reason" => reason}) do
+    {conn
+     |> redirect(to: conn.request_path)
```

Comment:
This is redirecting to the current page? Why can't we show the error here? Why do we need to store it on the server.